### PR TITLE
Prefer iterator over multiple `Vec::push(..)` calls

### PR DESCRIPTION
**Pull request by NathanSWard**
# Objective

- There are many uses of `Vec<T>` or similar containers that build one via a `for` loop and repeatedly calling `.push(T)`.
- Calling `.push(..)` in a loop is never ideal since every call requires the implementation to check if the `Vec` needs to allocate more space. This is particularity inefficient when the desired size of the resulting `Vec` is known.
- An often way to possibly optimize this is using `Vec::with_capacity`. This is helpful as the `Vec` being appended will never need to grow (an expensive operation). However, the `push` calls still have a branch where the `Vec`'s capacity is compared to its size.

## Solution

- In the case where a `Vec` is being created via iterating over a separate collection of elements and mapping these to a new result, use `source.map(..).collect(..)`. If the `source` has an exact size, the `Vec` implementation can bypass the need to check it's internal storage's size each time a new element is pushed. Also, the `Vec` can pre-allocate the necessary buffer size.
- There are multiple cases where many `Vec`'s are being constructed from a single iterable. In this case, `unzip` can be used.
- In the case where the iterator could possibly fail and short circuit via returning an error, if the desired size of the `Vec` is known, use `Vec::with_capacity` instead of just `Vec::new`.
- Finally, in the case were we have a pre-existing `Vec` with elements already inserted, prefer to use `extend(..)` to multiple calls of `push`. This has the same benefit if the provided iterator has an exact size. Only one check of the internal buffer's size is needed to check for reallocation instead of each individual call to `push`.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Good find. Let me file an issue for a clippy lint for this...

**Comment by SkiFire13** [comment]
> In this case, unzip can be used. However, this std::'s implementation only allows for 2 collections to be created.

You can use `unzip` to unzip an iterator of `(A, (B, C))` into a `(Vec<A>, (Vec<B>, Vec<C>))` and so on, thanks to `(impl Extend<A>, impl Extend<B>)` implementing `Extend<(A, B)>` (see [this impl](https://doc.rust-lang.org/stable/std/iter/trait.Extend.html#impl-Extend%3C(A%2C%20B)%3E))

**Comment by rparrett** [comment]
This PR seems to break just about all rendering with runtime panics like
(running the sprite example)
```
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `sprite_pipeline`
    vertex buffer 0 stride 3 does not respect `VERTEX_STRIDE_ALIGNMENT`
```

File: crates/bevy_render/src/mesh/shape/torus.rs [hunk 1]
```diff
@@ -50,39 +53,28 @@ impl From<Torus> for Mesh {
...
+            .flat_map(|(segment, side)| {
+                let lt = (side + segment * n_vertices_per_row) as u32;
+                let rt = ((side + 1) + segment * n_vertices_per_row) as u32;
+
+                let lb = (side + (segment + 1) * n_vertices_per_row) as u32;
+                let rb = ((side + 1) + (segment + 1) * n_vertices_per_row) as u32;
+
+                [lt, rt, lb, rt, rb, lb]
+            })
+            .collect();
```
[the remark this answers is not in the document]
Comment (NathanSWard) [reply crates/bevy_render/src/mesh/shape/torus.rs:77]: Looking at the [assembly](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=6b36ddb1fe9e146c89b8a5b49e59b968) it seems they're roughly the same. The `cartesian_product` version actually has a slightly smaller assembly size (not that smaller asm is always faster though).

[the remark this answers is not in the document]
Comment (SkiFire13) [reply crates/bevy_render/src/mesh/shape/torus.rs:77]: Oh I see, the `flat_map` loses information on the length of the total iterator and that precludes any stdlib's internal optimization.

[the remark this answers is not in the document]
Comment (NathanSWard) [reply crates/bevy_render/src/mesh/shape/torus.rs:77]: Quite unfortunate :(

**Comment by superdump** [comment]
Got any benchmarks of median system execution times from Tracy for systems of interest?

File: crates/bevy_render/src/camera/camera.rs [hunk 2]
```diff
@@ -350,10 +350,9 @@ pub fn camera_system<T: CameraProjection + Component>(
         })
         .collect();
 
-    let mut added_cameras = vec![];
-    for entity in &queries.p1() {
-        added_cameras.push(entity);
-    }
+    #[allow(clippy::needless_collect)]
```
Comment (cart) [inline crates/bevy_render/src/camera/camera.rs:353]: Whats the deal with this needless collect? False positive? Seems necessary to me.

File: crates/bevy_render/src/mesh/shape/torus.rs [hunk 3]
```diff
@@ -27,18 +28,20 @@ impl From<Torus> for Mesh {
...
+        let positions: Vec<[f32; 3]>;
+        let normals: Vec<[f32; 3]>;
+        let uvs: Vec<[f32; 2]>;
+
+        (positions, (normals, uvs)) = (0..=torus.subdivisions_segments)
+            .map(|segment| {
+                let theta = segment_stride * segment as f32;
+                (segment, theta)
+            })
+            .cartesian_product(0..=torus.subdivisions_sides)
```
Comment (cart) [inline crates/bevy_render/src/mesh/shape/torus.rs:43]: We've avoided the itertools dep in our tree for this long. I'd prefer to keep it it that way if this is the only justification.

**Review by cart — COMMENTED** [review]
I think I'm on board for most of these, especially for most cases that would benefit from having a pre-allocated vec for an exact size iterator.

I personally think for loops often read better (and align more naturally with peoples' programming instincts), especially in cases where the point of the for loop isn't one specific output. Ex: when iterating `gltf.materials()` in `loader.rs`, the point isn't _just_ to convert the materials to equivalent bevy materials. Its also to populate named_materials, linear_textures, etc. For-looping leaves the focus on "iterating the collection and producing arbitrary side effects", whereas mapping puts the focus on generating one _specific_ side effect. Iterator chaining often tends to result in more indentation and newlines (as evidenced by the increased line counts).

_That_ being said, theres nothing here that I would explicitly ask you to undo. Most of these are close to neutral refactors (sometimes slightly less legible, sometimes slightly more legible) that increase the likelihood of preallocating + removing checks.

File: crates/bevy_render/src/camera/camera.rs (hunk 2 above)
Comment (devil-ira) [reply crates/bevy_render/src/camera/camera.rs:353]: `needless_collect` is known to cause quite a few false positives.

**Comment by richchurcher** [comment]
Backlog cleanup: closed with tracking issue #15660 due to inactivity and likely significant bit rot over the last few years.

**Pull request closed without merging**